File: src/services/characters/chapterEventService.ts
/**
 * AI Novel Studio - 章节事件服务
 * 桌面端以 SQLite 为权威；浏览器开发模式才使用 localStorage。
 */
import { dbCall, lsGet, lsSet, isTauri, generateId, nowISO } from '../database/db';
import type {
  ChapterEvent,
  ChapterEventSource,
  ChapterEventStatus,
  CreateChapterEventInput,
} from '../../types/chapterEvent';

const KEY = 'ai_novel_studio_chapter_events';

interface ChapterEventDto {
  id?: unknown;
  novelId?: unknown;
  novel_id?: unknown;
  chapterId?: unknown;
  chapter_id?: unknown;
  title?: unknown;
  description?: unknown;
  involvedCharacterIds?: unknown;
  involved_character_ids?: unknown;
  impact?: unknown;
  risk?: unknown;
  status?: unknown;
  source?: unknown;
  aiTaskId?: unknown;
  ai_task_id?: unknown;
  createdAt?: unknown;
  created_at?: unknown;
  updatedAt?: unknown;
  updated_at?: unknown;
}

function getAll(): ChapterEvent[] {
  return lsGet<ChapterEvent[]>(KEY) ?? [];
}
function saveAll(items: ChapterEvent[]): void {
  lsSet(KEY, items);
}

function firstString(...values: unknown[]): string | undefined {
  return values.find((value): value is string => typeof value === 'string');
}

function requiredString(field: string, ...values: unknown[]): string {
  const value = firstString(...values);
  if (value === undefined) throw new Error(`章节事件数据缺少 ${field}`);
  return value;
}

function parseCharacterIds(value: unknown): string[] | undefined {
  if (Array.isArray(value)) {
    return value.filter((item): item is string => typeof item === 'string');
  }
  if (typeof value !== 'string' || !value.trim()) return undefined;
  try {
    const parsed = JSON.parse(value) as unknown;
    return Array.isArray(parsed)
      ? parsed.filter((item): item is string => typeof item === 'string')
      : undefined;
  } catch {
    return undefined;
  }
}

function isStatus(value: unknown): value is ChapterEventStatus {
  return (
    value === 'candidate' ||
    value === 'selected' ||
    value === 'required' ||
    value === 'forbidden' ||
    value === 'adopted' ||
    value === 'discarded'
  );
}

function isSource(value: unknown): value is ChapterEventSource {
  return value === 'manual' || value === 'ai_suggested';
}

function mapEvent(dto: ChapterEventDto): ChapterEvent {
  const status = dto.status;
  const source = dto.source;
  return {
    id: requiredString('id', dto.id),
    novelId: requiredString('novelId', dto.novelId, dto.novel_id),
    chapterId: requiredString('chapterId', dto.chapterId, dto.chapter_id),
    title: requiredString('title', dto.title),
    description: firstString(dto.description) ?? '',
    involvedCharacterIds: parseCharacterIds(dto.involvedCharacterIds ?? dto.involved_character_ids),
    impact: firstString(dto.impact),
    risk: firstString(dto.risk),
    status: isStatus(status) ? status : 'candidate',
    source: isSource(source) ? source : 'manual',
    aiTaskId: firstString(dto.aiTaskId, dto.ai_task_id),
    createdAt: requiredString('createdAt', dto.createdAt, dto.created_at),
    updatedAt: requiredString('updatedAt', dto.updatedAt, dto.updated_at),
  };
}

export const chapterEventService = {
  async getByChapterId(chapterId: string): Promise<ChapterEvent[]> {
    if (!chapterId) return [];
    if (isTauri()) {
      const list = await dbCall<ChapterEventDto[]>('list_chapter_events', { chapterId });
      return (list ?? []).map(mapEvent);
    }
    return getAll().filter((event) => event.chapterId === chapterId);
  },
  async create(input: CreateChapterEventInput): Promise<ChapterEvent> {
    if (isTauri()) {
      const dto = await dbCall<ChapterEventDto>('create_chapter_event', {
        input: {
          novelId: input.novelId,
          chapterId: input.chapterId,
          title: input.title,
          description: input.description,
          involvedCharacterIds: input.involvedCharacterIds,
          impact: input.impact,
          risk: input.risk,
          status: input.status,
          source: input.source,
        },
      });
      return mapEvent(dto);
    }
    const list = getAll();
    const now = nowISO();
    const ev: ChapterEvent = {
      ...input,
      id: generateId(),
      status: input.status || 'candidate',
      source: input.source || 'manual',
      createdAt: now,
      updatedAt: now,
    };
    list.push(ev);
    saveAll(list);
    return ev;
  },
  async update(id: string, input: Partial<ChapterEvent>): Promise<ChapterEvent | null> {
    if (isTauri()) {
      try {
        const dto = await dbCall<ChapterEventDto>('update_chapter_event', {
          id,
          input: {
            title: input.title,
            description: input.description,
            involvedCharacterIds: input.involvedCharacterIds,
            impact: input.impact,
            risk: input.risk,
            status: input.status,
          },
        });
        return mapEvent(dto);
      } catch {
        return null;
      }
    }
    const list = getAll();
    const idx = list.findIndex((e) => e.id === id);
    if (idx === -1) return null;
    list[idx] = { ...list[idx], ...input, updatedAt: nowISO() };
    saveAll(list);
    return list[idx];
  },
  async setStatus(id: string, status: ChapterEventStatus): Promise<void> {
    if (isTauri()) {
      await dbCall('set_chapter_event_status', { id, status });
      return;
    }
    const list = getAll();
    const idx = list.findIndex((e) => e.id === id);
    if (idx === -1) return;
    list[idx].status = status;
    list[idx].updatedAt = nowISO();
    saveAll(list);
  },
  async remove(id: string): Promise<void> {
    if (isTauri()) {
      await dbCall('delete_chapter_event', { id });
      return;
    }
    saveAll(getAll().filter((e) => e.id !== id));
  },
};

File: src-tauri/src/services/chapter_event_service.rs
use crate::domain::world::{ChapterEventDto, CreateChapterEventInput, UpdateChapterEventInput};
use crate::repositories::chapter_event_repository;
use rusqlite::Connection;

pub fn chapter_event_status(value: Option<&str>, fallback: &str) -> Result<String, String> {
    let status = value.unwrap_or(fallback).trim();
    match status {
        "candidate" | "selected" | "required" | "forbidden" | "adopted" | "discarded" => {
            Ok(status.to_string())
        }
        _ => Err("章节事件状态无效".to_string()),
    }
}

pub fn encode_character_ids(ids: Option<&[String]>) -> Option<String> {
    ids.filter(|values| !values.is_empty())
        .and_then(|values| serde_json::to_string(values).ok())
}

pub fn list_chapter_events(
    conn: &Connection,
    chapter_id: &str,
) -> Result<Vec<ChapterEventDto>, String> {
    chapter_event_repository::find_chapter_events_by_chapter(conn, chapter_id)
}

pub fn get_chapter_event(conn: &Connection, id: &str) -> Result<Option<ChapterEventDto>, String> {
    chapter_event_repository::find_chapter_event_by_id(conn, id)
}

pub fn create_chapter_event(
    conn: &Connection,
    input: CreateChapterEventInput,
) -> Result<ChapterEventDto, String> {
    let title = input.title.trim();
    if title.is_empty() || title.chars().count() > 240 {
        return Err("事件标题无效".to_string());
    }
    let status = chapter_event_status(input.status.as_deref(), "candidate")?;
    let source = match input.source.as_deref().unwrap_or("manual") {
        "manual" | "ai_suggested" => input.source.unwrap_or_else(|| "manual".to_string()),
        _ => return Err("事件来源无效".to_string()),
    };
    let id = uuid::Uuid::new_v4().to_string();
    let now = chrono::Utc::now().to_rfc3339();
    let involved = encode_character_ids(input.involved_character_ids.as_deref());

    chapter_event_repository::insert_chapter_event(
        conn,
        &id,
        &input.novel_id,
        &input.chapter_id,
        title,
        &input.description,
        involved.as_deref(),
        input.impact.as_deref(),
        input.risk.as_deref(),
        &status,
        &source,
        input.ai_task_id.as_deref(),
        &now,
    )?;

    chapter_event_repository::find_chapter_event_by_id(conn, &id)?
        .ok_or_else(|| "无法读取创建后的章节事件".to_string())
}

pub fn update_chapter_event(
    conn: &Connection,
    id: &str,
    input: UpdateChapterEventInput,
) -> Result<ChapterEventDto, String> {
    let now = chrono::Utc::now().to_rfc3339();
    let status = match &input.status {
        Some(value) => Some(chapter_event_status(Some(value), "candidate")?),
        None => None,
    };
    let involved = encode_character_ids(input.involved_character_ids.as_deref());

    chapter_event_repository::update_chapter_event_fields(
        conn,
        id,
        input.title.as_deref(),
        input.description.as_deref(),
        involved.as_deref(),
        input.impact.as_deref(),
        input.risk.as_deref(),
        status.as_deref(),
        &now,
    )?;

    chapter_event_repository::find_chapter_event_by_id(conn, id)?
        .ok_or_else(|| "无法读取更新后的章节事件".to_string())
}

pub fn set_chapter_event_status(conn: &Connection, id: &str, status: &str) -> Result<(), String> {
    let valid_status = chapter_event_status(Some(status), "candidate")?;
    let now = chrono::Utc::now().to_rfc3339();
    chapter_event_repository::update_chapter_event_status(conn, id, &valid_status, &now)
}

pub fn delete_chapter_event(conn: &Connection, id: &str) -> Result<(), String> {
    chapter_event_repository::delete_chapter_event(conn, id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::Connection;

    fn setup_test_db() -> Connection {
        let mut conn = Connection::open_in_memory().unwrap();
        crate::db::create_tables(&mut conn).unwrap();
        conn.execute(
            "INSERT INTO novels (id, title, created_at, updated_at) VALUES ('novel-1', '测试小说', '2026-01-01T00:00:00Z', '2026-01-01T00:00:00Z')",
            [],
        ).unwrap();
        conn.execute(
            "INSERT INTO chapters (id, novel_id, title, order_index, status, word_count, created_at, updated_at) VALUES ('chapter-1', 'novel-1', '第一章 觉醒', 1, 'not_started', 0, '2026-01-01T00:00:00Z', '2026-01-01T00:00:00Z')",
            [],
        ).unwrap();
        conn
    }

    #[test]
    fn test_chapter_event_lifecycle() {
        let conn = setup_test_db();
        let event = create_chapter_event(
            &conn,
            CreateChapterEventInput {
                novel_id: "novel-1".to_string(),
                chapter_id: "chapter-1".to_string(),
                title: "遭遇怪兽袭击".to_string(),
                description: "在荒野区遭遇铁毛猪攻击，惊险脱困。".to_string(),
                involved_character_ids: Some(vec!["char-1".to_string()]),
                impact: Some("获得怪兽材料".to_string()),
                risk: Some("受轻伤".to_string()),
                status: Some("candidate".to_string()),
                source: Some("manual".to_string()),
                ai_task_id: None,
            },
        )
        .unwrap();

        assert_eq!(event.title, "遭遇怪兽袭击");
        assert_eq!(event.status, "candidate");

        set_chapter_event_status(&conn, &event.id, "adopted").unwrap();
        let fetched = get_chapter_event(&conn, &event.id).unwrap().unwrap();
        assert_eq!(fetched.status, "adopted");

        let list = list_chapter_events(&conn, "chapter-1").unwrap();
        assert_eq!(list.len(), 1);

        delete_chapter_event(&conn, &event.id).unwrap();
        let list_after = list_chapter_events(&conn, "chapter-1").unwrap();
        assert_eq!(list_after.len(), 0);
    }
}
